src/AI/CommanderProductionInfo.java

```diff
@@ -24,47 +25,50 @@
 public class CommanderProductionInfo
 {
   public Army myArmy;
-  public Set<UnitModel> availableUnitModels;
+  public Set<ModelForCO> availableUnitModels;
   public Set<MapLocation> availableProperties;
-  public Map<Terrain.TerrainType, Integer> propertyCounts;
+  public Map<Commander, Map<Terrain.TerrainType, Integer>> propertyCounts;
   public Map<UnitModel, Set<TerrainType>> modelToTerrainMap;
 
   /**
    * Build a model of the production capabilities for a given Commander.
    * Could be used for your own, or your opponent's.
    */
-  public CommanderProductionInfo(Army co, GameMap gameMap, boolean includeFriendlyOccupied)
+  public CommanderProductionInfo(Army army, GameMap gameMap, boolean includeFriendlyOccupied)
   {
     // Figure out what unit types we can purchase with our available properties.
-    myArmy = co;
-    availableUnitModels = new HashSet<UnitModel>();
-    availableProperties = new HashSet<MapLocation>();
-    propertyCounts = new HashMap<Terrain.TerrainType, Integer>();
-    modelToTerrainMap = new HashMap<UnitModel, Set<TerrainType>>();
+    myArmy = army;
+    availableUnitModels = new HashSet<>();
+    availableProperties = new HashSet<>();
+    propertyCounts = new HashMap<>();
+    for( Commander co : army.cos )
+      propertyCounts.put(co, new HashMap<>());
+    modelToTerrainMap = new HashMap<>();
```

Comment:
Why remove all the type parameters?

Reply:
I get that Java can infer them based on the declarations above, but is there any reason to prefer not having them there? Or did you just decide you wanted fewer characters to look at?

Reply:
I like fewer characters, and it means fewer places to update if I ever want to change the types.